### PR TITLE
optimizations

### PR DESCRIPTION
**Pull request by mrdcvlsc**
### Don't forget -O2 or -O3 flags when compiling!!!

-------------------------------------------------

### Changes

- optimized ```InvMixColumns()```
- micro-optimized ```MixColumns()``` (needs benchmark test since the performance is so close with the old version)
- optimized ```ShiftRows()```
- optimized ```InvShiftRows()```
- removed the unused functions due to the new optimized code
- added an ```unsigned char *RoundedKey;``` member in AES class
- added one type of overload for each modes of AES encryption/decryption [ECB,CBC,CFB] where the ***key expansion*** is not done inside, this will provide faster performance when you are only using the same one **key** when encrypting/decrypting **different** vectors/arrays of unsigned char, because you don't need to compute the _roundKeys_ over and over again.
- added an overload for ```KeyExpansion()``` where it takes a **key** to generate the round key values of the member ```unsigned char *RoundedKey;```

Example use of the added overload methods

```c++
std::vector<unsigned char> plain = { 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff };
unsigned char key[] = { 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f };

AES aes(128);
aes.KeyExpansion(key); <--- Generate the ```roundKey``` once, that will be used in new overloads for encrypting&decrypting

unsigned char *out = aes.EncryptECB(plain.data(), plain.size(), len); <-- no need to pass the key, no keyExpansion
unsigned char *decrypted = aes.DecryptECB(out,len); <-- no need to pass the key, no keyExpansion

delete [] out;
delete [] decrypted;
```

### Test Output

```shell
[==========] Running 25 tests from 4 test suites.
[----------] Global test environment set-up.
[----------] 3 tests from KeyLengths
[ RUN      ] KeyLengths.KeyLength128
[       OK ] KeyLengths.KeyLength128 (0 ms)
[ RUN      ] KeyLengths.KeyLength192
[       OK ] KeyLengths.KeyLength192 (0 ms)
[ RUN      ] KeyLengths.KeyLength256
[       OK ] KeyLengths.KeyLength256 (0 ms)
[----------] 3 tests from KeyLengths (0 ms total)

[----------] 10 tests from ECB
[ RUN      ] ECB.EncryptDecrypt
[       OK ] ECB.EncryptDecrypt (0 ms)
[ RUN      ] ECB.EncryptDecryptVector
[       OK ] ECB.EncryptDecryptVector (0 ms)
[ RUN      ] ECB.OneBlockEncrypt
[       OK ] ECB.OneBlockEncrypt (0 ms)
[ RUN      ] ECB.OneBlockEncryptVector
[       OK ] ECB.OneBlockEncryptVector (0 ms)
[ RUN      ] ECB.OneBlockWithoutByteEncrypt
[       OK ] ECB.OneBlockWithoutByteEncrypt (0 ms)
[ RUN      ] ECB.OneBlockPlusOneByteEncrypt
[       OK ] ECB.OneBlockPlusOneByteEncrypt (0 ms)
[ RUN      ] ECB.TwoBlocksEncrypt
[       OK ] ECB.TwoBlocksEncrypt (0 ms)
[ RUN      ] ECB.OneBlockDecrypt
[       OK ] ECB.OneBlockDecrypt (0 ms)
[ RUN      ] ECB.OneBlockDecryptVector
[       OK ] ECB.OneBlockDecryptVector (0 ms)
[ RUN      ] ECB.TwoBlocksDecrypt
[       OK ] ECB.TwoBlocksDecrypt (0 ms)
[----------] 10 tests from ECB (1 ms total)

[----------] 6 tests from CBC
[ RUN      ] CBC.EncryptDecrypt
[       OK ] CBC.EncryptDecrypt (0 ms)
[ RUN      ] CBC.EncryptDecryptVector
[       OK ] CBC.EncryptDecryptVector (0 ms)
[ RUN      ] CBC.TwoBlocksEncrypt
[       OK ] CBC.TwoBlocksEncrypt (0 ms)
[ RUN      ] CBC.TwoBlocksEncryptVector
[       OK ] CBC.TwoBlocksEncryptVector (0 ms)
[ RUN      ] CBC.TwoBlocksDecrypt
[       OK ] CBC.TwoBlocksDecrypt (0 ms)
[ RUN      ] CBC.TwoBlocksDecryptVector
[       OK ] CBC.TwoBlocksDecryptVector (0 ms)
[----------] 6 tests from CBC (0 ms total)

[----------] 6 tests from CFB
[ RUN      ] CFB.EncryptDecrypt
[       OK ] CFB.EncryptDecrypt (0 ms)
[ RUN      ] CFB.EncryptDecryptVector
[       OK ] CFB.EncryptDecryptVector (0 ms)
[ RUN      ] CFB.EncryptTwoBlocks
[       OK ] CFB.EncryptTwoBlocks (0 ms)
[ RUN      ] CFB.EncryptTwoBlocksVector
[       OK ] CFB.EncryptTwoBlocksVector (0 ms)
[ RUN      ] CFB.DecryptTwoBlocks
[       OK ] CFB.DecryptTwoBlocks (0 ms)
[ RUN      ] CFB.DecryptTwoBlocksVector
[       OK ] CFB.DecryptTwoBlocksVector (0 ms)
[----------] 6 tests from CFB (0 ms total)

[----------] Global test environment tear-down
[==========] 25 tests from 4 test suites ran. (3 ms total)
[  PASSED  ] 25 tests.
```

### CODE REVIEW

**Comment by SergeyBel** [comment]
* optimized InvMixColumns() and micro-optimized MixColumns() - may be useful
* optimized ShiftRows()and optimized InvShiftRows() - this "optimization" has no noticeable improvement effect, but it greatly reduces readability of code. it does more harm than good
* added one type of overload for each modes of AES...- maybe idea makes sense but implementation is not good. Calls like `aes.KeyExpansion(key);` very inconvenient and the code itself is in bad style: a lot of code duplication and etc. I think it's not really needed yet

Moreover:
* Please don't modify Makefile and dev "infrastructure" except in obvious situations like typos
* There are a lot of useless comments like `<- please don't put this on heap...`  - they are not useful, only distracting

To summarize:
optimization of InvMixColumns() and MixColumns() looks interesting (I have not analyzed it in detail) and can be proposed in a separate PR. Rest changes are not needed
